biogeophys/FatesHydroWTFMod.F90

```diff
@@ -0,0 +1,1136 @@
+    write(fates_log(),*) 'check how the class pointer was setup'
+    call endrun(msg=errMsg(sourcefile, __LINE__))
+  end function th_from_psi_base
+  function psi_from_th_base(this,th) result(psi)
+    class(wrf_type)     :: this
+    real(r8),intent(in) :: th
+    real(r8)            :: psi
+    write(fates_log(),*) 'The base water retention function'
+    write(fates_log(),*) 'should never be actualized'
+    write(fates_log(),*) 'check how the class pointer was setup'
+    call endrun(msg=errMsg(sourcefile, __LINE__))
+  end function psi_from_th_base
+  function dpsidth_from_th_base(this,th) result(dpsidth)
+    class(wrf_type)     :: this
+    real(r8),intent(in) :: th
+    real(r8)            :: dpsidth
+    write(fates_log(),*) 'The base water retention function'
+    write(fates_log(),*) 'should never be actualized'
+    write(fates_log(),*) 'check how the class pointer was setup'
+    call endrun(msg=errMsg(sourcefile, __LINE__))
+  end function dpsidth_from_th_base
+  function ftc_from_psi_base(this,psi) result(ftc)
+    class(wkf_type)     :: this
+    real(r8),intent(in) :: psi
+    real(r8)            :: ftc
+    write(fates_log(),*) 'The base water retention function'
+    write(fates_log(),*) 'should never be actualized'
+    write(fates_log(),*) 'check how the class pointer was setup'
+    call endrun(msg=errMsg(sourcefile, __LINE__))
+  end function ftc_from_psi_base
+  function dftcdpsi_from_psi_base(this,psi) result(dftcdpsi)
+    class(wkf_type)     :: this
+    real(r8),intent(in) :: psi
+    real(r8)            :: dftcdpsi
+    write(fates_log(),*) 'The base water retention function'
+    write(fates_log(),*) 'should never be actualized'
+    write(fates_log(),*) 'check how the class pointer was setup'
+    call endrun(msg=errMsg(sourcefile, __LINE__))
+  end function dftcdpsi_from_psi_base
+
+  ! =====================================================================================
+  ! Van Genuchten Functions are defined here
+  ! =====================================================================================
+
+  subroutine set_wrf_param_vg(this,params_in)
+
+    class(wrf_type_vg) :: this
+    real(r8), intent(in) :: params_in(:)
+
+    this%alpha  = params_in(1)
+    this%psd    = params_in(2)
+    this%th_sat = params_in(3)
+    this%th_res = params_in(4)
+
+    return
+  end subroutine set_wrf_param_vg
+
+  ! =====================================================================================
+
+  subroutine set_wkf_param_vg(this,params_in)
+
+    class(wkf_type_vg) :: this
+    real(r8), intent(in) :: params_in(:)
+
+    this%alpha  = params_in(1)
+    this%psd    = params_in(2)
+    this%th_sat = params_in(3)
+    this%th_res = params_in(4)
+    this%tort   = params_in(5)
+
+    return
+  end subroutine set_wkf_param_vg
+
+  ! =====================================================================================
+
+  function th_from_psi_vg(this,psi) result(th)
```

Comment:
I had originally intended to allow positive pressures in all the different functional hypotheses, not just the tfs style. If I can add the linear extrapolations to VG and CCH, will do, otherwise, I will add a failure clause.

Reply:
@rgknox , Sounds good to me.